# IBM Applied Data Science Capstone Project

## Week 5 Final Assignment Notebook

### Opening a New Restaurant in Hyderabad, India.

- Build a dataframe of neighborhoods in Hyderabad, India by web scraping the data from Wikipedia page


- Get the geographical coordinates of the neighborhoods


- Obtain the venue data for the neighborhoods from Foursquare API


- Explore and cluster the neighborhoods


- Select the best cluster to open a new Restaurant

### Import libraries:

In [1]:
!pip install geocoder
!pip install folium

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import geocoder # to get coordinates

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print("Libraries imported.")

Libraries imported.


### Scrap data from Wikipedia page into a DataFrame

In [3]:
# send the GET request
data = requests.get("https://en.wikipedia.org/wiki/Category:Neighbourhoods_in_Hyderabad,_India").text

In [4]:
#data

In [5]:
# parse data from the html into a beautifulsoup object
soup = BeautifulSoup(data, 'html.parser')

In [6]:
# create a list to store neighborhood data
neighborhoodList = []

In [7]:
# append the data into the list
for row in soup.find_all("div", class_="mw-category")[0].findAll("li"):
    neighborhoodList.append(row.text)

In [8]:
# create a new DataFrame from the list
hy_df = pd.DataFrame({"Neighborhood": neighborhoodList})

hy_df.head()

Neighborhood
0     A. C. Guards
1  A. S. Rao Nagar
2  Abhyudaya Nagar
3            Abids
4         Adibatla

This is the data frame created after scraping the data. We need to get the geographical coordinates in the form of latitude and longitude in order to be able to use Foursquare API. To do so, we will use the Geocoder package that will allow us to convert the address into geographical coordinates in the form of latitude and longitude.

In [9]:
# Defining a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Hyderabad, India'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [10]:
# Call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(neighborhood) for neighborhood in hy_df["Neighborhood"].tolist()]

We have obtained the latitude and longitude coordinates for all the places, so we need to merge the coordinates into the original data frame.

In [11]:
# Create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [12]:
# Merge the coordinates into the original dataframe
hy_df['Latitude'] = df_coords['Latitude']
hy_df['Longitude'] = df_coords['Longitude']

In [15]:
hy_df.head()

Neighborhood   Latitude  Longitude
0     A. C. Guards  17.395015  78.459812
1  A. S. Rao Nagar  17.411200  78.508240
2  Abhyudaya Nagar  17.337650  78.564140
3            Abids  17.389800  78.476580
4         Adibatla  17.235790  78.541300

In [16]:
hy_df.shape

(200, 3)

This is the combined data frame which contains all the neighbourhoods along with the geographical coordinates.

In [18]:
# Getting the coordinates of Hyderabad
address = 'Hyderabad, India'
geolocator = Nominatim(user_agent="Hyderabad's mapping app", timeout=2)
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Hyderabad, India {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Hyderabad, India 17.360589, 78.4740613.


After gathering the data, we have to populate the data into a pandas DataFrame and then visualize the neighbourhoods in a map using Folium package.

In [76]:
map_hy = folium.Map(location=[latitude, longitude], zoom_start=11)
# Adding markers to map
for lat, lng, neighborhood in zip(hy_df['Latitude'],  hy_df['Longitude'], hy_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker([lat, lng],radius=5,popup=label,color='blue',fill=True,fill_color='#3186cc',fill_opacity=0.7).add_to(map_hy)
map_hy

In [22]:
# save the map as HTML file
map_hy.save(r"C:\Users\Hp\Desktop\P-jupyter\Others")

### Use the foursquare API to explore the neighbourhoods

In [23]:
CLIENT_ID = '2X1KKPYCKZ3COAHGCHROCBG2OWG5GO1JLS4L5DCVUSCMRTXI' # your Foursquare ID
CLIENT_SECRET = 'M5GUUMI0OV3LG31YVHMPENRYW4JTQ5X1PGMJI2DLKGBKWHCX' # your Foursquare Secret
VERSION = '20180604'
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 2X1KKPYCKZ3COAHGCHROCBG2OWG5GO1JLS4L5DCVUSCMRTXI
CLIENT_SECRET:M5GUUMI0OV3LG31YVHMPENRYW4JTQ5X1PGMJI2DLKGBKWHCX


In [25]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(hy_df['Latitude'], hy_df['Longitude'], hy_df['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

After extracting all the venues, we have to convert the venues list into a new DataFrame.

In [26]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head(10)

(6258, 7)


Neighborhood   Latitude  Longitude                           VenueName  \
0  A. C. Guards  17.395015  78.459812              Cafe Niloufer & Bakers   
1  A. C. Guards  17.395015  78.459812                       Subhan Bakery   
2  A. C. Guards  17.395015  78.459812                            Chicha's   
3  A. C. Guards  17.395015  78.459812                          Nizam club   
4  A. C. Guards  17.395015  78.459812        Taiba Bakers & Confectioners   
5  A. C. Guards  17.395015  78.459812                 Rayalaseema Ruchulu   
6  A. C. Guards  17.395015  78.459812                   Shahi Dastarkhwan   
7  A. C. Guards  17.395015  78.459812                     Laxman Ki Bandi   
8  A. C. Guards  17.395015  78.459812  Siraj's International Juice Center   
9  A. C. Guards  17.395015  78.459812                             Spice 6   

   VenueLatitude  VenueLongitude            VenueCategory  
0      17.399715       78.462881                     Café  
1      17.392412       78.464712                   Bakery  
2      17.403255       78.460152    Hyderabadi Restaurant  
3      17.403221       78.468729                   Lounge  
4      17.402530       78.456823                   Bakery  
5      17.403084       78.463012  South Indian Restaurant  
6      17.402640       78.461399    Hyderabadi Restaurant  
7      17.378895       78.463973  South Indian Restaurant  
8      17.395579       78.442701                Juice Bar  
9      17.409007       78.450559                   Bistro

### Lets check how many venues were returned for each neighbourhood

In [29]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  \
Neighborhood                                                           
A. C. Guards                                70         70         70   
A. S. Rao Nagar                             20         20         20   
Abhyudaya Nagar                              9          9          9   
Abids                                       79         79         79   
Adikmet                                     19         19         19   
Afzal Gunj                                  46         46         46   
Aghapura                                    65         65         65   
Aliabad, Hyderabad                           8          8          8   
Alijah Kotla                                16         16         16   
Allwyn Colony                               16         16         16   
Alwal                                        6          6          6   
Amberpet                                    17         17         17   
Ameenpur                                     2          2          2   
Ameerpet                                   100        100        100   
Anandbagh                                   11         11         11   
Ashok Nagar, Hyderabad                      69         69         69   
Asif Nagar                                  13         13         13   
Attapur                                      6          6          6   
Azamabad, Hyderabad                         41         41         41   
Azampura                                    22         22         22   
BHEL Township, Hyderabad                     5          5          5   
Badichowdi                                  65         65         65   
Bagh Lingampally                            45         45         45   
Bairamalguda                                 5          5          5   
Balapur, Ranga Reddy district                1          1          1   
Balkampet                                   48         48         48   
Banjara Hills                              100        100        100   
Bank Street, Hyderabad                      79         79         79   
Barkas, Hyderabad                            6          6          6   
Barkatpura                                  46         46         46   
Basheerbagh                                 98         98         98   
Bazarghat                                   69         69         69   
Begum Bazaar                                29         29         29   
Begumpet                                    52         52         52   
Bharat Nagar                                 4          4          4   
Boggulkunta                                 80         80         80   
Bolarum                                      5          5          5   
Borabanda                                   15         15         15   
Bowenpally                                  24         24         24   
Chaderghat                                  38         38         38   
Champapet                                    5          5          5   
Chanchalguda                                14         14         14   
Chandrayan Gutta                            10         10         10   
Chatta Bazaar                               27         27         27   
Chengicherla                                 1          1          1   
Cherlapally                                  8          8          8   
Chikkadpally                                60         60         60   
Chilkalguda                                 20         20         20   
Chintal Basti                              100        100        100   
Chintalakunta                                4          4          4   
Dabirpura                                   19         19         19   
Dar-ul-Shifa                                32         32         32   
Dhoolpet                                    12         12         12   
Dilsukhnagar                                17         17         17   
Domalg

### Let's find out how many unique categories can be curated from all the returned venues

In [30]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 175 uniques categories.


In [32]:
venues_df['VenueCategory'].unique()[:50] #displays all the 175 category names

array(['Café', 'Bakery', 'Hyderabadi Restaurant', 'Lounge',
       'South Indian Restaurant', 'Juice Bar', 'Bistro', 'Science Museum',
       'Ice Cream Shop', 'Indian Restaurant', 'Park',
       'Middle Eastern Restaurant', 'Stadium', 'Performing Arts Venue',
       'Dessert Shop', 'Vegetarian / Vegan Restaurant', 'Hotel Bar',
       'Hotel', 'Shoe Store', 'Neighborhood', 'Diner', 'Snack Place',
       'Coffee Shop', 'Pizza Place', 'Mobile Phone Shop',
       'Fast Food Restaurant', 'Chinese Restaurant', 'Clothing Store',
       'Fried Chicken Joint', 'Department Store', 'Shopping Mall',
       'Soccer Field', 'Electronics Store', 'Hookah Bar',
       'Breakfast Spot', 'Planetarium', 'Movie Theater',
       'Convenience Store', 'Sandwich Place', 'Asian Restaurant',
       'Fruit & Vegetable Store', 'Restaurant', 'Food Truck',
       'Chaat Place', 'Burger Joint', 'Bar', 'Smoke Shop', 'Food',
       'Multiplex', 'Gaming Cafe'], dtype=object)

In [34]:
# check if the results contain "Restaurant"
"Restaurant" in venues_df['VenueCategory'].unique()

True

### Analyze each neighbourhood

Here we apply one hot encoding to all the venues. So now the number of columns becomes 175

In [35]:
# one hot encoding
hy_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
hy_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [hy_onehot.columns[-1]] + list(hy_onehot.columns[:-1])
hy_onehot = hy_onehot[fixed_columns]

print(hy_onehot.shape)
hy_onehot.head(20)

(6258, 176)


Neighborhoods  ATM  Afghan Restaurant  Airport Service  \
0   A. C. Guards    0                  0                0   
1   A. C. Guards    0                  0                0   
2   A. C. Guards    0                  0                0   
3   A. C. Guards    0                  0                0   
4   A. C. Guards    0                  0                0   
5   A. C. Guards    0                  0                0   
6   A. C. Guards    0                  0                0   
7   A. C. Guards    0                  0                0   
8   A. C. Guards    0                  0                0   
9   A. C. Guards    0                  0                0   
10  A. C. Guards    0                  0                0   
11  A. C. Guards    0                  0                0   
12  A. C. Guards    0                  0                0   
13  A. C. Guards    0                  0                0   
14  A. C. Guards    0                  0                0   
15  A. C. Guards    0                  0                0   
16  A. C. Guards    0                  0                0   
17  A. C. Guards    0                  0                0   
18  A. C. Guards    0                  0                0   
19  A. C. Guards    0                  0                0   

    American Restaurant  Andhra Restaurant  Arcade  Arts & Crafts Store  \
0                     0                  0       0                    0   
1                     0                  0       0                    0   
2                     0                  0       0                    0   
3                     0                  0       0                    0   
4                     0                  0       0                    0   
5                     0                  0       0                    0   
6                     0                  0       0                    0   
7                     0                  0       0                    0   
8                     0                  0       0                    0   
9                     0                  0       0                    0   
10                    0                  0       0                    0   
11                    0                  0       0                    0   
12                    0                  0       0                    0   
13                    0                  0       0                    0   
14                    0                  0       0                    0   
15                    0                  0       0                    0   
16                    0                  0       0                    0   
17                    0                  0       0                    0   
18                    0                  0       0                    0   
19                    0                  0       0                    0   

    Asian Restaurant  Athletics & Sports  Auditorium  BBQ Joint  \
0                  0                   0           0          0   
1                  0                   0           0          0   
2                  0                   0           0          0   
3                  0                   0           0          0   
4                  0                   0           0          0   
5                  0                   0           0          0   
6                  0                   0           0          0   
7                  0                   0           0          0   
8                  0                   0           0          0   
9                  0                   0           0          0   
10                 0                   0           0          0   
11                 0                   0           0          0   
12                 0                   0           0          0   
13                 0                   0           0          0   
14                 0                   0           0          0   
15                 0                   0           0          0   
16    

Next, let’s group rows of neighbourhood by taking the sum of the frequency of occurrence of each category.

In [39]:
hy_grouped=hy_onehot.groupby(["Neighborhoods"]).sum().reset_index()

hy_grouped

Neighborhoods  ATM  Afghan Restaurant  \
0                            A. C. Guards    0                  0   
1                         A. S. Rao Nagar    0                  0   
2                         Abhyudaya Nagar    0                  0   
3                                   Abids    0                  0   
4                                 Adikmet    0                  0   
5                              Afzal Gunj    0                  0   
6                                Aghapura    0                  0   
7                      Aliabad, Hyderabad    0                  0   
8                            Alijah Kotla    0                  0   
9                           Allwyn Colony    0                  0   
10                                  Alwal    0                  0   
11                               Amberpet    0                  0   
12                               Ameenpur    1                  0   
13                               Ameerpet    0                  0   
14                              Anandbagh    0                  0   
15                 Ashok Nagar, Hyderabad    0                  0   
16                             Asif Nagar    0                  0   
17                                Attapur    0                  0   
18                    Azamabad, Hyderabad    0                  0   
19                               Azampura    0                  0   
20               BHEL Township, Hyderabad    0                  0   
21                             Badichowdi    0                  0   
22                       Bagh Lingampally    0                  0   
23                           Bairamalguda    0                  0   
24          Balapur, Ranga Reddy district    0                  0   
25                              Balkampet    0                  0   
26                          Banjara Hills    0                  0   
27                 Bank Street, Hyderabad    0                  0   
28                      Barkas, Hyderabad    0                  1   
29                             Barkatpura    0                  0   
30                            Basheerbagh    0                  0   
31                              Bazarghat    0                  0   
32                           Begum Bazaar    0                  0   
33                               Begumpet    0                  0   
34                           Bharat Nagar    2                  0   
35                            Boggulkunta    0                  0   
36                                Bolarum    0                  0   
37                              Borabanda    0                  0   
38                             Bowenpally    0                  0   
39                             Chaderghat    0                  0   
40                              Champapet    0                  0   
41                           Chanchalguda    0                  0   
42                       Chandrayan Gutta    0                  1   
43                          Chatta Bazaar    0                  0   
44                           Chengicherla    1                  0   
45                            Cherlapally    4                  0   
46                           Chikkadpally    0                  0   
47                            Chilkalguda    0                  0   
48                          Chintal Basti    0                  0   
49                          Chintalakunta    0                  0   
50                              Dabirpura    0                  0   
51                           Dar-ul-Shifa    0                  0   
52                               Dhoolpet    0                  0   
53                           Dilsukhnagar    0                  0   
54                              Domalguda    0                  0   
55                           ECIL X Roads    0                  0   
56                              Edi Bazar    2                  0   
57                        

In [41]:
print(hy_grouped.shape)

(197, 176)


In [42]:
rest = hy_grouped.loc[:, hy_grouped.columns.str.endswith('Restaurant')]
rest

Afghan Restaurant  American Restaurant  Andhra Restaurant  \
0                    0                    0                  0   
1                    0                    0                  0   
2                    0                    0                  0   
3                    0                    0                  0   
4                    0                    0                  0   
5                    0                    0                  0   
6                    0                    0                  0   
7                    0                    0                  0   
8                    0                    0                  0   
9                    0                    0                  0   
10                   0                    0                  0   
11                   0                    0                  0   
12                   0                    0                  0   
13                   0                    1                  1   
14                   0                    0                  0   
15                   0                    0                  0   
16                   0                    0                  0   
17                   0                    0                  0   
18                   0                    0                  0   
19                   0                    0                  0   
20                   0                    0                  0   
21                   0                    0                  0   
22                   0                    0                  0   
23                   0                    0                  0   
24                   0                    0                  0   
25                   0                    0                  0   
26                   0                    2                  0   
27                   0                    0                  0   
28                   1                    0                  0   
29                   0                    0                  0   
30                   0                    0                  0   
31                   0                    0                  0   
32                   0                    0                  0   
33                   0                    0                  1   
34                   0                    0                  0   
35                   0                    0                  0   
36                   0                    0                  0   
37                   0                    0                  0   
38                   0                    0                  0   
39                   0                    0                  0   
40                   0                    0                  0   
41                   0                    0                  0   
42                   1                    0                  0   
43                   0                    0                  0   
44                   0                    0                  0   
45                   0                    0                  0   
46                   0                    0                  0   
47                   0                    0                  0   
48                   0                    1                  0   
49                   0                    0                  0   
50                   0                    0                  0   
51                   0                    0                  0   
52                   0                    0                  0   
53                   0                    0                  0   
54                   0                    0                  0   
55                   0                    0                  0   
56                   0                    0                  0   
57                   0                    0                  0   
58                   0                    0                  0   
59                   0                    0  

In [47]:
rest.shape

(197, 33)

In [56]:
rest['Total_Restaurants']=rest.iloc[:,-33:].sum(axis=1)

rest

<ipython-input-56-8b95490b16a7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rest['Total_Restaurants']=rest.iloc[:,-33:].sum(axis=1)


Afghan Restaurant  American Restaurant  Andhra Restaurant  \
0                    0                    0                  0   
1                    0                    0                  0   
2                    0                    0                  0   
3                    0                    0                  0   
4                    0                    0                  0   
5                    0                    0                  0   
6                    0                    0                  0   
7                    0                    0                  0   
8                    0                    0                  0   
9                    0                    0                  0   
10                   0                    0                  0   
11                   0                    0                  0   
12                   0                    0                  0   
13                   0                    1                  1   
14                   0                    0                  0   
15                   0                    0                  0   
16                   0                    0                  0   
17                   0                    0                  0   
18                   0                    0                  0   
19                   0                    0                  0   
20                   0                    0                  0   
21                   0                    0                  0   
22                   0                    0                  0   
23                   0                    0                  0   
24                   0                    0                  0   
25                   0                    0                  0   
26                   0                    2                  0   
27                   0                    0                  0   
28                   1                    0                  0   
29                   0                    0                  0   
30                   0                    0                  0   
31                   0                    0                  0   
32                   0                    0                  0   
33                   0                    0                  1   
34                   0                    0                  0   
35                   0                    0                  0   
36                   0                    0                  0   
37                   0                    0                  0   
38                   0                    0                  0   
39                   0                    0                  0   
40                   0                    0                  0   
41                   0                    0                  0   
42                   1                    0                  0   
43                   0                    0                  0   
44                   0                    0                  0   
45                   0                    0                  0   
46                   0                    0                  0   
47                   0                    0                  0   
48                   0                    1                  0   
49                   0                    0                  0   
50                   0                    0                  0   
51                   0                    0                  0   
52                   0                    0                  0   
53                   0                    0                  0   
54                   0                    0                  0   
55                   0                    0                  0   
56                   0                    0                  0   
57                   0                    0                  0   
58                   0                    0                  0   
59                   0                    0  

Join the total restaurants column to the grouped dataframe.

In [57]:
extracted_col = rest["Total_Restaurants"]
hy_grouped = hy_grouped.join(extracted_col)
hy_grouped.head()

Neighborhoods  ATM  Afghan Restaurant  Airport Service  \
0     A. C. Guards    0                  0                0   
1  A. S. Rao Nagar    0                  0                0   
2  Abhyudaya Nagar    0                  0                0   
3            Abids    0                  0                0   
4          Adikmet    0                  0                0   

   American Restaurant  Andhra Restaurant  Arcade  Arts & Crafts Store  \
0                    0                  0       0                    0   
1                    0                  0       0                    0   
2                    0                  0       0                    0   
3                    0                  0       0                    0   
4                    0                  0       0                    0   

   Asian Restaurant  Athletics & Sports  Auditorium  BBQ Joint  \
0                 0                   0           0          0   
1                 1                   0           0          0   
2                 0                   0           0          0   
3                 0                   0           0          0   
4                 0                   0           0          0   

   Badminton Court  Bakery  Bank  Bar  Baseball Field  Bed & Breakfast  \
0                0       5     0    0               0                0   
1                0       0     0    0               0                0   
2                0       0     0    0               0                0   
3                0       5     0    1               0                0   
4                0       1     0    0               0                0   

   Beer Garden  Bengali Restaurant  Big Box Store  Bike Rental / Bike Share  \
0            0                   0              0                         0   
1            0                   0              0                         0   
2            0                   0              0                         0   
3            0                   0              0                         0   
4            0                   0              0                         0   

   Bistro  Bookstore  Bowling Alley  Breakfast Spot  Brewery  Buffet  \
0       1          0              0               1        0       0   
1       0          0              0               0        0       0   
2       0          0              0               0        0       0   
3       0          0              0               1        0       0   
4       0          0              0               0        0       0   

   Burger Joint  Bus Station  Business Service  Cable Car  Cafeteria  Café  \
0             0            0                 0          0          0     3   
1             0            0                 0          0          0     2   
2             0            0                 0          0          0     0   
3             1            0                 0          0          0     3   
4             0            0                 0          0          0     2   

   Candy Store  Castle  Chaat Place  Cheese Shop  Chinese Restaurant  \
0            0       0            0            0                   1   
1            0       0            0            0                   0   
2            0       0            0            0                   0   
3            0       0            1            0                   4   
4            0       0            0            0                   0   

   Clothing Store  Cocktail Bar  Coffee Shop  Comfort Food Restaurant  \
0               1             0            1                        0   
1               0             0            2                        0   
2               0             0            1                        0   
3               1             0            2                        0   
4               0             0            2                        0   

   Concert Hall  Convenience Store  Convention Center  Cricket Ground  \
0             0  

### Total number of restaurants in all neighbourhoods in Hyderabad

In [60]:
Total = hy_grouped['Total_Restaurants'].sum()
Total

74708

- So there are 74708 restaurants in hyderabad(which it very high)

- So now we want to select a good location where the no of restaurants are less so that our chances of setting up a restaurant at that location  should be good

## Create a dataframe for shopping mall data only

In [61]:
hy_res = hy_grouped[["Neighborhoods","Total_Restaurants"]]
hy_res.head()

Neighborhoods  Total_Restaurants
0     A. C. Guards                908
1  A. S. Rao Nagar                307
2  Abhyudaya Nagar                 80
3            Abids                920
4          Adikmet                268

## Now cluster the neighbourhoods

Run k-means to cluster the neighborhoods in Hyderabad into 3 clusters

In [66]:
# set number of clusters
kclusters = 4

hy_clustering = hy_res.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(hy_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]


array([2, 3, 0, 2, 3, 3, 2, 0, 0, 3], dtype=int32)

In [67]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
hy_merged = hy_res.copy()

# add clustering labels
hy_merged["Cluster Labels"] = kmeans.labels_

In [68]:
hy_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
hy_merged.head(10)

Neighborhood  Total_Restaurants  Cluster Labels
0        A. C. Guards                908               2
1     A. S. Rao Nagar                307               3
2     Abhyudaya Nagar                 80               0
3               Abids                920               2
4             Adikmet                268               3
5          Afzal Gunj                360               3
6            Aghapura                828               2
7  Aliabad, Hyderabad                120               0
8        Alijah Kotla                 80               0
9       Allwyn Colony                280               3

In [69]:
#Add latitude and longitude values by using the join operation(the new dataframe with the old dataframe containing the latitude and longitude values)
#hy_merged = hy_merged.join(hy_df.set_index("Neighborhood"), on="Neighborhood")

#print(hy_merged.shape)
hy_merged['Latitude'] = df_coords['Latitude']
hy_merged['Longitude'] = df_coords['Longitude']
hy_merged.head() # check the last columns!

Neighborhood  Total_Restaurants  Cluster Labels   Latitude  Longitude
0     A. C. Guards                908               2  17.395015  78.459812
1  A. S. Rao Nagar                307               3  17.411200  78.508240
2  Abhyudaya Nagar                 80               0  17.337650  78.564140
3            Abids                920               2  17.389800  78.476580
4          Adikmet                268               3  17.235790  78.541300

In [71]:
# sorting the results by Cluster Labels
print(hy_merged.shape)
hy_merged.sort_values(["Cluster Labels"], inplace=True)
hy_merged

(197, 5)


Neighborhood  Total_Restaurants  Cluster Labels  \
196                           Shivam Road                 80               0   
123                              Mallapur                  0               0   
86                         Kakatiya Nagar                 80               0   
153                     Nawab Saheb Kunta                 80               0   
83                             Jeedimetla                  0               0   
82                          Jamia Osmania                188               0   
81                       Jalal Baba Nagar                  0               0   
80                               Jahanuma                  0               0   
79                         Jagadgirigutta                  0               0   
77                           IDA Bollaram                  0               0   
74                            Hayathnagar                  0               0   
73                           Hastinapuram                 40               0   
88                                  Kapra                160               0   
72                             Hasmathpet                 80               0   
156                              Nizampet                  0               0   
68                      Gundlapochampalli                 40               0   
67                            Gulzar Houz                 80               0   
158                          Old Neredmet                200               0   
129         Mehboob ki Mehendi, Hyderabad                120               0   
163                            Patancheru                  0               0   
63                          Gautham Nagar                  0               0   
62                           Gaddiannaram                148               0   
136                            Moghalpura                 80               0   
59                              Ferozguda                 40               0   
58                 Fateh Nagar, Hyderabad                160               0   
155                              Neredmet                 80               0   
57                              Erragadda                 80               0   
122                     Malkajgiri mandal                 80               0   
121                            Malkajgiri                 80               0   
137                              Moosapet                  0               0   
117            Madina building, Hyderabad                120               0   
115                           Madeenaguda                120               0   
114                            Madannapet                204               0   
113                        Macha Bollaram                 40               0   
112                             Lothkunta                 40               0   
138                          Moosarambagh                 80               0   
110                           Langar Houz                  0               0   
109                             Lallaguda                120               0   
108                           Lal Darwaza                 40               0   
139                 Moti Nagar, Hyderabad                 40               0   
90                             Karmanghat                 39               0   
105                           Laad Bazaar                 80               0   
103                            Kushaiguda                  0               0   
102                            Kukatpally                200               0   
140                             Moula Ali                200               0   
143                              Nacharam                 40               0   
99                    Kothapet, Hyderabad                 80               0   
144  Nagaram, Medchal–Malkajgiri district                120               0   
97                               Kompally                160               0   
96             Kings Colony, 

In [72]:
hy_merged['Total_Restaurants'].max()

1862

## Now we visualize the resulting clusters

In [77]:

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(hy_merged['Latitude'], hy_merged['Longitude'], hy_merged['Neighborhood'], hy_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Cluster 0

In [75]:
hy_merged.loc[hy_merged['Cluster Labels'] == 0]

Neighborhood  Total_Restaurants  Cluster Labels  \
196                           Shivam Road                 80               0   
123                              Mallapur                  0               0   
86                         Kakatiya Nagar                 80               0   
153                     Nawab Saheb Kunta                 80               0   
83                             Jeedimetla                  0               0   
82                          Jamia Osmania                188               0   
81                       Jalal Baba Nagar                  0               0   
80                               Jahanuma                  0               0   
79                         Jagadgirigutta                  0               0   
77                           IDA Bollaram                  0               0   
74                            Hayathnagar                  0               0   
73                           Hastinapuram                 40               0   
88                                  Kapra                160               0   
72                             Hasmathpet                 80               0   
156                              Nizampet                  0               0   
68                      Gundlapochampalli                 40               0   
67                            Gulzar Houz                 80               0   
158                          Old Neredmet                200               0   
129         Mehboob ki Mehendi, Hyderabad                120               0   
163                            Patancheru                  0               0   
63                          Gautham Nagar                  0               0   
62                           Gaddiannaram                148               0   
136                            Moghalpura                 80               0   
59                              Ferozguda                 40               0   
58                 Fateh Nagar, Hyderabad                160               0   
155                              Neredmet                 80               0   
57                              Erragadda                 80               0   
122                     Malkajgiri mandal                 80               0   
121                            Malkajgiri                 80               0   
137                              Moosapet                  0               0   
117            Madina building, Hyderabad                120               0   
115                           Madeenaguda                120               0   
114                            Madannapet                204               0   
113                        Macha Bollaram                 40               0   
112                             Lothkunta                 40               0   
138                          Moosarambagh                 80               0   
110                           Langar Houz                  0               0   
109                             Lallaguda                120               0   
108                           Lal Darwaza                 40               0   
139                 Moti Nagar, Hyderabad                 40               0   
90                             Karmanghat                 39               0   
105                           Laad Bazaar                 80               0   
103                            Kushaiguda                  0               0   
102                            Kukatpally                200               0   
140                             Moula Ali                200               0   
143                              Nacharam                 40               0   
99                    Kothapet, Hyderabad                 80               0   
144  Nagaram, Medchal–Malkajgiri district                120               0   
97                               Kompally                160               0   
96             Kings Colony, 

In [80]:
hy_merged.loc[hy_merged['Cluster Labels'] == 0].sum()

Neighborhood         Shivam RoadMallapurKakatiya NagarNawab Saheb K...
Total_Restaurants                                                 8045
Cluster Labels                                                       0
Latitude                                                   1741.473702
Longitude                                                  7847.864063
dtype: object

#### There are 8045 restaurants in the cluster 0 

### Cluster 1

In [81]:
hy_merged.loc[hy_merged['Cluster Labels'] == 1]

Neighborhood  Total_Restaurants  Cluster Labels   Latitude  \
61                Gachibowli               1503               1  17.474111   
166              Pisal Banda               1268               1  17.526770   
116                 Madhapur               1862               1  17.525910   
48             Chintal Basti               1094               1  17.426750   
168               Punjagutta               1347               1  17.524900   
127               Masab Tank               1374               1  17.391860   
84             Jubilee Hills               1305               1  17.421967   
69                HITEC City               1678               1  17.363380   
13                  Ameerpet               1415               1  17.533320   
174          Raj Bhavan Road               1347               1  17.505370   
185                 Saifabad               1188               1  17.328500   
75   Himayatnagar, Hyderabad               1147               1  17.327100   
76                 Hyderguda               1147               1  17.404860   
93                Khairtabad               1214               1  17.379070   
154                  Nayapul               1268               1  17.394430   
26             Banjara Hills               1221               1  17.415460   
179     Red Hills, Hyderabad               1068               1  17.411200   
173                  Raidurg               1107               1  17.382373   
30               Basheerbagh               1188               1  17.402110   
107             Lakdi ka pul               1336               1  17.361450   
54                 Domalguda               1078               1  17.368560   

     Longitude  
61   78.426392  
166  78.252340  
116  78.376330  
48   78.512290  
168  78.449580  
127  78.459050  
84   78.525592  
69   78.475020  
13   78.325290  
174  78.467490  
185  78.538340  
75   78.605310  
76   78.481210  
93   78.436680  
154  78.427900  
26   78.434330  
179  78.508240  
173  78.482083  
30   78.477700  
107  78.472090  
54   78.535220

In [82]:
hy_merged.loc[hy_merged['Cluster Labels'] == 1].sum()

Neighborhood         GachibowliPisal BandaMadhapurChintal BastiPunj...
Total_Restaurants                                                27155
Cluster Labels                                                      21
Latitude                                                    365.869451
Longitude                                                  1647.668476
dtype: object

#### There are 27155 restaurants in the cluster 1 

### Cluster 2

In [83]:
hy_merged.loc[hy_merged['Cluster Labels'] == 2]

Neighborhood  Total_Restaurants  Cluster Labels  \
161                           Pan bazar                815               2   
151                          Nanalnagar                878               2   
152                         Narayanguda                959               2   
164                               Patny                896               2   
178                           Rani Gunj                815               2   
188                        Sanghi Nagar                896               2   
189                        Santoshnagar                746               2   
191                        Secunderabad                776               2   
149                 Nampally, Hyderabad                868               2   
132            Minister Road, Hyderabad                775               2   
0                          A. C. Guards                908               2   
130                         Mehdipatnam                986               2   
3                                 Abids                920               2   
6                              Aghapura                828               2   
15               Ashok Nagar, Hyderabad                854               2   
25                            Balkampet                684               2   
27               Bank Street, Hyderabad                880               2   
31                            Bazarghat                828               2   
33                             Begumpet                830               2   
35                          Boggulkunta                880               2   
46                         Chikkadpally                587               2   
78                          Izzat Nagar                639               2   
85                            Kachiguda                588               2   
60                           Film Nagar                998               2   
98                             Kondapur               1039               2   
101            Krishna Nagar, Hyderabad                956               2   
124                          Mallepally                748               2   
119  Mahatma Gandhi Road (Secunderabad)                815               2   

      Latitude  Longitude  
161  17.462590  78.516300  
151  17.423410  78.515340  
152  17.389310  78.467450  
164  17.433790  78.492220  
178  17.390320  78.545440  
188  17.405390  78.471380  
189  17.477175  78.528480  
191  17.396330  78.457048  
149  17.375379  78.498487  
132  17.362310  78.471027  
0    17.395015  78.459812  
130  17.401390  78.453540  
3    17.389800  78.476580  
6    17.377510  78.480050  
15   17.457870  78.538820  
25   17.448530  78.449250  
27   17.388601  78.476645  
31   17.393236  78.460698  
33   17.447290  78.453960  
35   17.505990  78.304540  
46   17.448288  78.576924  
78   17.399220  78.480750  
85   17.521830  78.454330  
60   17.458410  78.451810  
98   17.325980  78.447350  
101  17.368354  78.525035  
124  17.447370  78.535200  
119  17.456960  78.390050

In [84]:
hy_merged.loc[hy_merged['Cluster Labels'] == 2].sum()

Neighborhood         Pan bazarNanalnagarNarayangudaPatnyRani GunjSa...
Total_Restaurants                                                23392
Cluster Labels                                                      56
Latitude                                                    487.747648
Longitude                                                  2197.378517
dtype: object

#### There are 23392 restaurants in the cluster 2

### Cluster 3

In [85]:
hy_merged.loc[hy_merged['Cluster Labels'] == 3]

Neighborhood  Total_Restaurants  Cluster Labels   Latitude  \
106             Lab quarters                388               3  17.512650   
32              Begum Bazaar                240               3  17.373580   
111              Lingojiguda                280               3  17.440300   
177      Ramnagar, Hyderabad                307               3  17.511610   
29                Barkatpura                559               3  17.393870   
118              Maharajgunj                400               3  17.495380   
22          Bagh Lingampally                519               3  17.399310   
21                Badichowdi                560               3  17.388960   
19                  Azampura                228               3  17.407100   
18       Azamabad, Hyderabad                547               3  17.369170   
16                Asif Nagar                240               3  17.407840   
135      Moazzam Jahi Market                480               3  17.432860   
9              Allwyn Colony                280               3  17.360680   
5                 Afzal Gunj                360               3  17.410610   
190              Saroornagar                240               3  17.458760   
4                    Adikmet                268               3  17.235790   
192          Serilingampally                240               3  17.392100   
126              Marredpally                536               3  17.288640   
1            A. S. Rao Nagar                307               3  17.411200   
38                Bowenpally                360               3  17.452700   
39                Chaderghat                280               3  17.469241   
94                 Khajaguda                356               3  17.422620   
42          Chandrayan Gutta                278               3  17.369180   
92                Kavadiguda                398               3  17.340610   
147               Nallakunta                519               3  17.609930   
148              Namalagundu                240               3  17.378930   
91                    Karwan                240               3  17.458100   
150             Nanakramguda                239               3  17.397480   
89    Karkhana, Secunderabad                416               3  17.365610   
87   Kamala Nagar, Hyderabad                279               3  17.386880   
100          Koti, Hyderabad                560               3  17.466600   
142               Mylargadda                240               3  17.453992   
71                 Hafeezpet                239               3  17.417120   
70                 Habsiguda                268               3  17.581230   
157           Noorkhan Bazar                280               3  17.394870   
66              Gudimalkapur                320               3  17.391920   
159  Padmanabha Nagar Colony                439               3  17.518310   
160           Padmarao Nagar                240               3  17.371820   
65                Goshamahal                400               3  17.325280   
162               Parsigutta                335               3  17.391600   
64                   Golnaka                279               3  17.365070   
131                Mettuguda                228               3  17.329790   
53              Dilsukhnagar                239               3  17.374070   
51              Dar-ul-Shifa                280               3  17.366180   
47               Chilkalguda                280               3  17.403230   
141              Musheerabad                346               3  17.370770   
195             Shanker Mutt                279               3  17.482160   
170               Putlibowli                280               3  17.514250   

     Longitude  
106  78.441290  
32   78.469210  
111  78.527910  
177  78.294110  
29   78.498420  
118  78.333360  
22   78.499640  
21   78.486810  
19   78.502330  
18   78.436830  
16   7

In [86]:
hy_merged.loc[hy_merged['Cluster Labels'] == 3].sum()

Neighborhood         Lab quartersBegum BazaarLingojigudaRamnagar, H...
Total_Restaurants                                                16116
Cluster Labels                                                     144
Latitude                                                    835.859953
Longitude                                                  3766.731447
dtype: object

#### There are 16116 restaurants in the cluster 3

# Final observation

Major number of restaurants are concentrated in the central area of Hyderabad city, with the highest number in cluster 1 & 2  and moderate number in cluster 3. This represents a great opportunity and high potential areas to open new restaurants as there is very little to no competition from existing restaurants. Meanwhile, restaurants in cluster 2 are likely suffering from intense competition due to oversupply and high concentration of restaurants. Therefore, this project recommends property developers to capitalize on these findings to open new restaurants in neighbourhoods in cluster 0 with relatively less competition. Property developers with unique selling propositions to stand out from the competition can also open new restaurants in neighbourhoods in cluster 3 with moderate competition. Lastly, property developers are advised to prefer central area of hyderabad city with the recommended cluster because lot of demand can be seen and with best business model can yeild a best restaurant.